### ISSUE TITLE
Support for discriminated unions

### ISSUE THREAD

**Issue by alexrock** [microsoft/pyright#944]
First, thanks for this awesome project.

I would like to know if _discriminated unions_ will be supported in Pyright as they are in MyPy?

```py
from typing import Literal, cast

class A:
  kind: Literal['A']
  a: str

class B:
  kind: Literal['B']
  b: int

C = A | B

c = cast(C, A())

if c.kind == 'A':
  c.a # no error, c is A
else:
  c.b # no error, c is B
```

**Comment by erictraut** [comment]
Thanks for the suggestion. I've implemented this support, and it will be in the next release. I've added additional documentation [here](https://github.com/microsoft/pyright/blob/master/docs/internals.md#type-checking-concepts).

**Comment by alexrock** [comment]
Thank you so much

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.64, which I just published. It will also be included in the next release of Pylance.
